### PR TITLE
Fix sporadic unit test failure of ConvertSpiceDataToRealData 

### LINKED ISSUE #20520: Investigate sporadic build failure with ConvertCWPDMDToSpectraTest
This is one of the sporadic build failures on the August 2017 sheet.

### Expected behavior
Unit tests pass fine every time.

### Actual behavior
Occasionally it seems to fail due to the interplay between the `ConvertSpiceDataToRealSpice` algorithm run in the tests and the `DownloadInstrumentTest`.

### Steps to reproduce the behavior
I found I could make it fail sporadically like this:
```
for i in `seq 100`; do ctest -R "(ConvertSpice|DownloadInstrument)" -j8 --schedule-random --output-on-failure; done;
```

And I got the following output when it eventually fails:
```
ConvertSpiceDataToRealSpace-[Notice] ConvertSpiceDataToRealSpace started
ConvertSpiceDataToRealSpace-[Information] Algorithm: ConvertSpiceDataToRealSpace v1
ConvertSpiceDataToRealSpace-[Information] Execution Date: 2017-Sep-19 16:09:32
ConvertSpiceDataToRealSpace-[Information] Execution Duration: -1 seconds
ConvertSpiceDataToRealSpace-[Information] Parameters:
ConvertSpiceDataToRealSpace-[Information]   Name: InputWorkspace, Value: DataTable, Default?: No, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: RunInfoWorkspace, Value: LogParentWS, Default?: No, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: RunStart, Value: , Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: Instrument, Value: HB2A, Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: DetectorPrefix, Value: anode, Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: RunNumberName, Value: Pt., Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: RotationAngleLogName, Value: 2theta, Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: MonitorCountsLogName, Value: monitor, Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: DurationLogName, Value: time, Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information]   Name: OutputWorkspace, Value: HB2A_MD, Default?: No, Direction: Output
ConvertSpiceDataToRealSpace-[Information]   Name: OutputMonitorWorkspace, Value: MonitorMDW, Default?: No, Direction: Output
ConvertSpiceDataToRealSpace-[Information]   Name: DetectorEfficiencyTableWorkspace, Value: , Default?: Yes, Direction: Input
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Information] No parameter file found for this instrument
ConvertSpiceDataToRealSpace-[Information] Temporary workspace inherites run_start as 2012-08-13T11:57:33. It will be replaced by the correct value.
ExperimentInfo-[Information] run_start/start_time not stored in workspace. Default to current date.
LoadInstrument-[Error] Error in execution of algorithm LoadInstrument:
LoadInstrument-[Error] File not found
ConvertSpiceDataToRealSpace-[Error] Error in execution of algorithm ConvertSpiceDataToRealSpace:
ConvertSpiceDataToRealSpace-[Error] File not found
libc++abi.dylib: terminating with uncaught exception of type Poco::FileNotFoundException: File not found
Child aborted
```

### Platforms affected
All.

### PR DESCRIPTION
**Pull request by samueljackson92**
What was happening here is that the `DownloadInstrumentTest` creates a temporary folder in the instrument folder. If it's run at the same time as the `LoadInstrument` algorithm (which is run several times by `ConvertSpiceDataToRealData`) it can pick up the temporary test folder when parsing the IDF folder contents.

`Poco::File().isFile()` throws an exception if the C `stat` function fails for some reason (probably because the directory was subsequently removed). On the other hand `Poco::Path().isFile()` just returns `false` if it's not a file and we safely ignore the testing directory.

**To test:**
 - Build `DataHandlingTest` and `MDAlgorithmsTest`
 - This test only fails sporadically. I found I could get it to fail fairly quickly by running the following:
```
for i in `seq 10`; do ctest -R "(ConvertSpice|DownloadInstrument)" -j8 --schedule-random --output-on-failure; done;
``` 
 - Once you can reproduce, apply the fix then try and get it to fail again.
 - Check there are no other unit test failures.

Fixes #20520 

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by ElliotAOram — COMMENTED** [review]
Works on Ubuntu 16.04 (No failure observed in 50 repetitions)

**Review by ElliotAOram — APPROVED** [review]
Works on Windows 10 (No failure observed in 50 repetitions)

File: Framework/API/src/ExperimentInfo.cpp [hunk 1]
```diff
@@ -1004,7 +1004,8 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
     // find the first beat file
     for (Poco::DirectoryIterator dir_itr(directoryName); dir_itr != end_iter;
          ++dir_itr) {
-      if (!Poco::File(dir_itr->path()).isFile())
+
+      if (!Poco::Path(dir_itr->path()).isFile())
```
Comment (martyngigg) [inline Framework/API/src/ExperimentInfo.cpp]: Is this performing an unnecessary copy?

`dir_itr->path()` returns a `Poco::Path` object anyway so we could just do `if (!dir_itr->path().isFile())` ?

File: Framework/API/src/ExperimentInfo.cpp [hunk 2]
```diff
@@ -1004,7 +1004,8 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
...
         continue;
 
       std::string l_filenamePart = Poco::Path(dir_itr->path()).getFileName();
```
Comment (martyngigg) [inline Framework/API/src/ExperimentInfo.cpp]: I know you didn't change this one but I think we can remove the `Poco::Path()` part and just use the returned `Path` directly.

**Review by martyngigg — CHANGES_REQUESTED** [review]

File: Framework/API/src/ExperimentInfo.cpp (hunk 2 above)
Comment (samueljackson92) [reply Framework/API/src/ExperimentInfo.cpp]: I agree. I didn't look closely enough. I thought it was a `std::string`.

File: Framework/API/src/ExperimentInfo.cpp [hunk 3]
```diff
@@ -1004,15 +1004,18 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
     // find the first beat file
     for (Poco::DirectoryIterator dir_itr(directoryName); dir_itr != end_iter;
          ++dir_itr) {
-      if (!Poco::File(dir_itr->path()).isFile())
+
+      const auto filePath = dir_itr.path();
```
Comment (martyngigg) [inline Framework/API/src/ExperimentInfo.cpp]: `const auto &` would be good to avoid a copy.

File: Framework/API/src/ExperimentInfo.cpp [hunk 4]
```diff
@@ -1004,15 +1004,18 @@ ExperimentInfo::getInstrumentFilename(const std::string &instrumentName,
...
+      if (!filePath.isFile())
         continue;
 
-      std::string l_filenamePart = Poco::Path(dir_itr->path()).getFileName();
+      std::string l_filenamePart = filePath.getFileName();
       if (regex_match(l_filenamePart, regex)) {
-        g_log.debug() << "Found file: '" << dir_itr->path() << "'\n";
+        const auto pathName = dir_itr->path();
```
Comment (martyngigg) [inline Framework/API/src/ExperimentInfo.cpp]: Could we avoid the confusion with the `*` & `->` operators on `DirectoryIterator`? As we have the file now this should be the same as `filePath.path()`. Also a `const auto &` here would avoid a copy too.

**Review by martyngigg — CHANGES_REQUESTED** [review]

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
@mantidproject/gatekeepers Now that the release branch has been created could we merge this to hopefully improve the stability of the unit tests.

**Pull request merged**